core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,165 @@
+      case _ =>
+        val executorId = info.executorId
+        failedExecutorMap.get(executorId) match {
+          case Some(failureStatus) =>
+            failureStatus.updatedTime = clock.getTimeMillis()
+            val failedTimes = failureStatus.failedTasks.getOrElse(info.taskId, 0) + 1
+            failureStatus.failedTasks.update(info.taskId, failedTimes)
+          case None =>
+            val failedTasks = mutable.HashMap(info.taskId -> 1)
+            val failureStatus = new FailureStatus(
+              clock.getTimeMillis(),
+              info.host,
+              failedTasks)
+            failedExecutorMap.update(executorId, failureStatus)
+        }
+    }
+  }
+
+  def removeFailureExecutors(
+      executorId: String,
+      taskId: Option[Long]) : Unit = synchronized {
+        taskId match {
+          // If taskId is provided, remove the taskId from failedTasks
+          case Some(id) =>
+            failedExecutorMap.get(executorId).map(fs => {
+              fs.updatedTime = clock.getTimeMillis()
+              fs.failedTasks.remove(id)
+              if(fs.failedTasks.isEmpty){
+                failedExecutorMap.remove(executorId)
+              }
+            })
+          // if taskId is not provided, remove the whole record related to given executorId
+          case None =>
+            failedExecutorMap.remove(executorId)
+        }
+  }
+
+  def executorIsBlacklisted(
+      executorId: String,
+      sched: TaskSchedulerImpl,
+      taskId: Option[Long]) : Boolean = {
+
+    executorBlacklist(sched, taskId).contains(executorId)
+  }
+
+  // The actual implementation is delegated to strategy
+  def executorBlacklist(
+      sched: TaskSchedulerImpl,
+      taskId: Option[Long]): Set[String] = synchronized {
+
+    // If the node is in blacklist, all executors allocated on that node will
+    // also be put into  executor blacklist.
+    // By default it's turned off, user can enable it in sparkConf.
+    val speculationFailedExecutor: Set[String] =
+      if (sparkConf.getBoolean("spark.scheduler.blacklist.speculate", false)) {
+      Set.empty[String]
+    } else {
+      nodeBlacklist(taskId).flatMap(sched.getExecutorsAliveOnHost(_)
+          .getOrElse(Set.empty[String])).toSet
+    }
+
+    speculationFailedExecutor ++ strategy.getExecutorBlacklist(failedExecutorMap, taskId)
+  }
+
+  // The actual implementation is delegated to strategy
+  def nodeBlacklist(taskId: Option[Long] = None): Set[String] = synchronized {
+    strategy.getNodeBlacklist(failedExecutorMap, taskId)
+  }
+}
+
+/**
+ * A class to record details of failure.
+ *
+ * @param initialTime the time when failure status be created
+ * @param host the node name which running executor on
+ * @param failedTasks all tasks failed on the executor (key is task id,
+ *  value is failure time of this task)
+ */
+final class FailureStatus(
+    initialTime: Long,
+    val host: String,
+    val failedTasks: mutable.HashMap[Long, Int]) {
```

Comment:
probably just a stupid mistake on my part, but when I first read this I thought it was the (clock) time at which the failure occurred.  can we rename it to something like `numFailuresPerTask`?